runtime/compiler/p/codegen/J9PPCWatchedStaticFieldSnippet.cpp

```diff
@@ -0,0 +1,129 @@
+   str->method = TR::J9WatchedStaticFieldSnippet::staticFieldData.method;
+   str->location = TR::J9WatchedStaticFieldSnippet::staticFieldData.location;
+   str->fieldAddress = TR::J9WatchedStaticFieldSnippet::staticFieldData.fieldAddress;
+   str->fieldClass = TR::J9WatchedStaticFieldSnippet::staticFieldData.fieldClass;
+
+   if (cg()->comp()->getOption(TR_UseSymbolValidationManager))
+      {
+      cg()->addExternalRelocation(
+         new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, method), reinterpret_cast<uint8_t *>(staticFieldData.method), reinterpret_cast<uint8_t *>(TR::SymbolType::typeMethod), TR_SymbolFromManager, cg()),
+         __FILE__,
+         __LINE__,
+         node);
+      }
+   else
+      {
+      cg()->addExternalRelocation(new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, method), NULL, TR_RamMethod, cg()), __FILE__, __LINE__, node);
+      }
+
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   // If the field is unresolved then we populate these snippet fields with the correct value at runtime (via the instructions generated by generateFillInDataBlockSequenceForUnresolvedField)
+   // and hence don't need to add relocation records here.
+   if (isResolved)
+      {
+      cg()->addExternalRelocation(
+         new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, fieldAddress), reinterpret_cast<uint8_t *>(node->getSymbolReference()), reinterpret_cast<uint8_t *>(node->getInlinedSiteIndex()), TR_DataAddress, cg()),
+         __FILE__,
+         __LINE__,
+         node);
+      if (cg()->comp()->getOption(TR_UseSymbolValidationManager))
+         {
+         cg()->addExternalRelocation(
+               new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, fieldClass), reinterpret_cast<uint8_t *>(staticFieldData.fieldClass), reinterpret_cast<uint8_t *>(TR::SymbolType::typeClass), TR_SymbolFromManager, cg()),
+               __FILE__,
+               __LINE__,
+               node);
+         }
+      else
+         {
+         /* Deploy when AOT is Enabled. TR_ClassAddress needs to support contiguous addresses. 
+         TR_RelocationRecordInformation *recordInfo = (TR_RelocationRecordInformation *)cg()->comp()->trMemory()->allocateMemory(sizeof(TR_RelocationRecordInformation), heapAlloc);
+         recordInfo->data1 = (uintptr_t)(node->getSymbolReference());
+         recordInfo->data2 = (uintptr_t)(node->getInlinedSiteIndex());
+         recordInfo->data3 = (uintptr_t)(fixedSequence7);
+         cg()->addExternalRelocation(
+               new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, fieldClass), (uint8_t *)recordInfo, TR_ClassAddress, cg()),
+               __FILE__,
+               __LINE__,
+               node);
+         */
```

Comment:
Power needs to send additional information to identify that we're dealing with a contiguous TR_ClassAddress. When AOT is enabled, we would not be calling the base class's emitSnippetBody() for the static case, since that would lead to creating an invalid relocation. 

Overriding the emitSnippetBody is essential to successfully set the TR_ClassAddress relocation.

Reply:
@fjeremic Usually for a TR_ClassAddress the SymbolReference, and InlinedSiteIndex are passed as arguments. 
Unfortunately, the infrastructure doesn't support a third argument so instead P passes a TR_RelocationRecordInformation for some relocations to hold additional arguments. 

TR_ClassAddress is natively a discontiguous relocation. Passing this new fixedSequence allows us to treat the relocation as contiguous. 

Again, calling the base class' emitSnippetBody would create a discontiguous relocation, not contiguous.

Reply:
I think we need to take AOT support with FieldWatch on Power out of this PR. I will open a separate issue just for that. 

In this PR, we should put an ASSERT if we somehow get into this mode. Perhaps we can call the common emit method but put an ASSERT in the Power specific method.

Reply:
This only seems to affect "old AOT", i.e. in the default case where we use the symbol validation manager (SVM) we do not have this problem, is this correct?

Reply:
So in the "old AOT" we'd need to use TR_ClassAddress and pass in a new sequence number that means "no sequence" . In "new AOT" (ie SVM), we'd have to use TR_SymbolFromManager (since it's by default, contiguous).

Reply:
Overall, complicated coupling between how an address is patched and the relo type on POWER is extremely confusing and error-prone, and it's something I really want to get to as part of my long term clean up efforts.

Reply:
Opened separate issue: https://github.com/eclipse/openj9/issues/6256

Reply:
Note: it's definitely not part of 0.15 milestone.

Reply:
The original review has been fixed here:
https://github.com/eclipse/openj9/pull/6234/files#diff-1527bfb2af24a08e08921356946e547dR38

I think we can resolve this one.